### ISSUE TITLE
Distinction between browser and minimal-ui

### ISSUE THREAD

**Issue by frivoal** [w3c/manifest#1023]
There is no guidance on what the distinction between browser and minimal-ui should be.

* How is browser different from minimal-ui? The spec doesn't say.
* If minimal-ui and browser both have the same controls, do both values of `display-mode` match? What if app manifest was used to explicitly invoke one or the other?
* Is browser allowed to have fewer controls than minimal-ui? That seems somewhat unlikely in major products like Safari and Chrome, but the spec definition is loose enough that it seems allowed, if a platform's conventions for hyperlinks has less UI than what minimal-ui calls for.

**Comment by marcoscaceres** [comment]
> There is no guidance on what the distinction between browser and minimal-ui should be.
> How is browser different from minimal-ui? The spec doesn't say.

They are indeed similar... just to get a shared understanding, this has been our working model (apologies this was not clear):

 * "browser" is exactly the same behavior you get if you bookmark a web page: the web application is launched amongst other browser tabs/windows if the web browser is already open - or in a new tab, if the browser wasn't running. To the end-user, "browser" looks and feels exactly like any other web page/application and it's clear that application is running inside the a browser: "I can see the web page is running inside Firefox/Chrome/Safari/Edge". 

 * "minimal-ui": this launches in its own window but not with other browser tabs or identifiable/browser-branded UI  (i.e., it looks like a self contained web application). The UI that is provided is "minimal": like a back/forward/refresh buttons, maybe address bar that is "read-only" and no browser tabs. The user shouldn't be able to say, "it's running in < web browser name here >"... the application is fully "self contained".   

> If minimal-ui and browser both have the same controls, do both values of display-mode match? What if app manifest was used to explicitly invoke one or the other?

No. Browsers have things like tabs, often clear branding, bookmaking facilities, and other things that "minimal-ui" wouldn't have. Think of it like a bare-bones unbranded browser where you can only view this one web application.    

> Is browser allowed to have fewer controls than minimal-ui? That seems somewhat unlikely in major products like Safari and Chrome, but the spec definition is loose enough that it seems allowed, if a platform's conventions for hyperlinks has less UI than what minimal-ui calls for.

Browser would always have more controls (i.e., "it's a real browser"). Minimal UI as shown on latest Chrome/Android: 
![Screen Shot 2021-12-07 at 11 53 08 am](https://user-images.githubusercontent.com/870154/144946108-f3de9bcb-ea2d-414e-a7c8-1b97ad9c5ac4.png)

(Provides back in the hardware button, forward, reload, and other useful, but limited, things...)

Hopefully that makes more sense now... happy to coordinate in tightening up that text.

**Comment by marcoscaceres** [comment]
In case you have an Android device and want to check the above: 
https://wpt.live/appmanifest/display-member/display-member-media-feature-minimal-ui-manual.html

Note that minimal-ui is not supported in iOS.

**Comment by frivoal** [comment]
Thanks. I think the situation in the general case is pretty clear (at least in intent, I'm not sure the spec wording supports that intent fully). It's the edge cases where it gets a little fuzzy. I think the essence of the question hangs on this sentence of yours:

> Browsers have things like tabs, often clear branding, bookmaking facilities, and other things that "minimal-ui" wouldn't have

I'd rephrase it as:

> Browsers **may** have things like tabs, often clear branding, bookmaking facilities, and other things that "minimal-ui" wouldn't **typically** have.

And the the question is what do you do in the case of a simplistic browser, not meaningfully different from what you'd get from a `minimal-ui` UA? This isn't necessarily what we expect from a mainstream smartphone, but the concept of a display-mode is meant to be generic, and to apply to any sort of device with a web UA. Not sure the browser on every TV, set-top box, fridge, in-flight-navigation system, etc, is always that full featured.

Even if you just stick to desktop, have a look at the browser offered by https://servo.org/. Sure, it's an experimental browser, with hardly any focus at all on its UI, but it's a browser nonetheless, so it needs to answer that mediaquery somehow. I'd expect the `display-mode` media feature to match `browser`, since that's what it does in the typical case of a standalone browser. At the same time, if you actually look at its UI, it's pretty-much what you'd expect for `standalone` (I cannot find any other UI than CMD+L to open a URL). Should the query match both? Should it not match `browser` for some reason?

The query is explicitly defined as applying not only based on app manifest mode switching, but also to UAs that match the described mode. So I think we need a bit more.

**Comment by marcoscaceres** [comment]
> And the the question is what do you do in the case of a simplistic browser, not meaningfully different from what you'd get from a minimal-ui UA? This isn't necessarily what we expect from a mainstream smartphone, but the concept of a display-mode is meant to be generic, and to apply to any sort of device with a web UA. Not sure the browser on every TV, set-top box, fridge, in-flight-navigation system, etc, is always that full featured.

This is true. But at the same time, UAs are not required to support every display mode (or installable web apps at all). For instance, Firefox only supports installable web applications on mobile (and it doesn't support "minimal-ui"). 

> I'd expect the display-mode media feature to match browser, since that's what it does in the typical case of a standalone browser.

That's correct. 

>  At the same time, if you actually look at its UI, it's pretty-much what you'd expect for standalone (I cannot find any other UI than CMD+L to open a URL). Should the query match both? Should it not match browser for some reason?

It would match "browser". The only time it would match anything else is if the user agent: 

 1. The user agent has explicitly "[applied](https://www.w3.org/TR/appmanifest/#dfn-applied)" the manifest to the web page (i.e., the web manifest is affecting the presentation). 
 2. Supported some definition of "minimal-ui" or "standalone". 

In other words, because of 1, a regular web page would never match anything but "browser". 

> The query is explicitly defined as applying not only based on app manifest mode switching, but also to UAs that match the described mode. So I think we need a bit more.

Ok, if it says "but also to UAs that match the described mode" that's definitely a bug. It must only match when the manifest is [applied](https://www.w3.org/TR/appmanifest/#dfn-applied).

**Comment by frivoal** [comment]
> Ok, if it says "but also to UAs that match the described mode" that's definitely a bug. It must only match when the manifest is applied.

> …a regular web page would never match anything but "browser".

Hmmm… the spec looks very explicit to me that a regular web page should occasionally match things other than `browser` (second paragraph of [7.1](https://www.w3.org/TR/appmanifest/#the-display-mode-media-feature)):

> A user agent MUST expose the 'display-mode' media feature irrespective of whether a manifest is being applied to a browsing context. For example, if the end-user puts the whole user agent into fullscreen, then the user agent would reflect this change to CSS and scripts via the 'display-mode' media feature.

**Comment by marcoscaceres** [comment]
Ok, let me check if anyone actually implemented that. It's been a while (years ☺️) since I last checked.

**Comment by marcoscaceres** [comment]
Ah d'oh, you are right! "fullscreen" does indeed do fullscreen in Gecko, and Chrome (but not in Safari).

**Comment by marcoscaceres** [comment]
Ok, that's fun ... I guess it's back to better describing these :)

**Comment by marcoscaceres** [comment]
What about if "generally" the `"minimal-ui"` mapped to `window.open()` with [`windowFeatures`](https://developer.mozilla.org/en-US/docs/Web/API/Window/open#window_features) set to ["popup"](https://html.spec.whatwg.org/multipage/window-object.html#dom-open-dev). 

> "popup"
>  Encourages user agents to provide a minimal web browser user interface for the new window. (Impacts the visible getter on all BarProp objects as well.)

**Comment by dmurph** [comment]
I don't think this change is accurate - minimal-ui's point (AFAIK) was to have sufficient browser controls so the site doesn't have to implement their own back/reload navigation. popups don't have that?

**Comment by marcoscaceres** [comment]
Agree with @dmurph ... this display mode unique.

**Comment by christianliebel** [comment]
Ok, then I'm going to close this issue. I'm fine with the current definition, which already expresses that, and I believe it's accurate.

> This mode is similar to standalone, but provides the end-user with some means to access a minimal set of UI elements for controlling navigation (i.e., back, forward, reload, and perhaps some way of viewing the document's address).
